File: C++/SortingAlgorithms/CountingSort.cpp
#include "countingSort.h"


std::vector<unsigned int> counting_sort(std::vector<unsigned int> arr)
{
	int max = *std::max_element(arr.begin(), arr.end());

	std::vector<unsigned int> count(max + 1, 0);
	std::vector<unsigned int> output(arr.size());

	for (int num : arr) {
		count[num]++;
	}

	for (int i = 1; i <= max; i++) {
		count[i] += count[i - 1]; 
	}

	for (int j = arr.size() - 1; j >= 0; j--) {
		output[count[arr[j]] - 1] = arr[j];
		count[arr[j]]--;
	}
	return output;
}

File: Python/SortingAlgorithms/counting_sort.py
def counting_sort(arr):
    if not arr:
        return []

    max_val = max(arr)

    count = [0] * (max_val + 1)

    for n in arr:
        count[n]+=1

    for i in range(1, len(count)):
        count[i]+= count[i-1]

    output = [0]*len(arr)

    for num in reversed(arr):
        output[count[num] - 1] =num
        count[num] -=1

    return output







